python/hail/utils.py

```diff
@@ -52,3 +56,140 @@ def functions_rst(self, file_name):
 
 
+@handle_py4j
+def hadoop_read(path, buffer_size=8192):
+    """Open a readable file through the Hadoop filesystem API. 
+    Supports distributed file systems like hdfs, gs, and s3.
+    
+    **Example:**
+    
+    .. doctest::
+        :options: +SKIP
+
+        >>> with hadoop_read('gs://my-bucket/notes.txt') as f:
+        ...     for line in f:
+        ...         print(line.strip())
+    
+    .. note::
+        
+        If Spark is running in cluster mode, both the source and destination 
+        file paths must be URIs (uniform resource identifiers). This means 
+        fully clarified paths, prefixed by scheme (``file://``, ``hdfs://``, ``gs://``,
+        ``s3://``, etc.)
+
+    .. caution::
+    
+        These file handles are slower than standard Python file handles.
+        If you are reading a file larger than ~50M, it will be faster to 
+        use :py:meth:`~hail.hadoop_copy` to copy the file locally, then read it
+        with standard Python I/O tools.
+    
+    :param str path: Source file URI.
+    
+    :param int buffer_size: Size of internal buffer.
+    
+    :return: Iterable file reader.
+    :rtype: `io.BufferedReader <https://docs.python.org/2/library/io.html#io.BufferedReader>`_
+    """
+    if not isinstance(path, str) and not isinstance(path, unicode):
+        raise TypeError("expected parameter 'path' to be type str, but found %s" % type(path))
+    if not isinstance(buffer_size, int):
+        raise TypeError("expected parameter 'buffer_size' to be type int, but found %s" % type(buffer_size))
+    return io.BufferedReader(HadoopReader(path), buffer_size=buffer_size)
+
+@handle_py4j
+def hadoop_write(path, buffer_size=8192):
+    """Open a writable file through the Hadoop filesystem API. 
+    Supports distributed file systems like hdfs, gs, and s3.
+    
+    **Example:**
+    
+    .. doctest::
+        :options: +SKIP
+
+        >>> with hadoop_write('gs://my-bucket/notes.txt') as f:
+        ...     f.write('result1: %s\\n' % result1)
+        ...     f.write('result2: %s\\n' % result2)
+    
+    .. note::
+        
+        If Spark is running in cluster mode, both the source and destination 
+        file paths must be URIs (uniform resource identifiers). This means 
+        fully clarified paths, prefixed by scheme (``file://``, ``hdfs://``, ``gs://``,
+        ``s3://``, etc.)
+
+    .. caution::
+    
+        These file handles are slower than standard Python file handles. If you
+        are writing a large file (larger than ~50M), it will be faster to write
+        to a local file using standard Python I/O and use :py:meth:`~hail.hadoop_copy` 
+        to move your file to a distributed file system.
+    
+    :param str path: Destination file URI.
+    
+    :return: File writer object.
+    :rtype: `io.BufferedWriter <https://docs.python.org/2/library/io.html#io.BufferedWriter>`_
+    """
+    if not isinstance(path, str) and not isinstance(path, unicode):
```

Comment:
This is fine for now but we should probably have another conversation about parameter typechecking.